### ISSUE TITLE
[JENKINS-18407] Starting a job with params using REST Api doesn't return 201 http status code as stated in the documentation

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/jenkins#14661]
<p>This bug is related to the new enhancement that was added in v1.519 for better tracking a queued job.<br/>
When starting the job with parameters I expect to get 201 status code, but instead I'm getting 302 code which redirects me to the "Location" header url.<br/>
See attached screenshot from the request and the response (taken by Fiddler)</p>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/amoss166.png" title="amoss166's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=amoss166">amoss166</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-18407" target="_blank">Starting a job with params using REST Api doesn't return 201 http status code as stated in the documentation</a></i></summary>
<i><ul>
<li><b>assignee</b>: <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kohsuke.png" title="kohsuke's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kohsuke">kohsuke</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Critical
<li><b>component(s)</b>: core
<li><b>label(s)</b>: API, rest
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2018-10-01T13:05:50+00:00
<li><b>votes</b>: 6
<li><b>watchers</b>: 11
<li><b>imported</b>: 2025-11-24
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>This bug is related to the new enhancement that was added in v1.519 for better tracking a queued job.
When starting the job with parameters I expect to get 201 status code, but instead I'm getting 302 code which redirects me to the "Location" header url.
See attached screenshot from the request and the response (taken by Fiddler)</p></pre>
</details>
</details>
<ul><li><i>environment</i>: <code>Linux</code></li></ul>
<details><summary><i>1 attachment</i></summary>

- [restBug (1).JPG](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-core-cli/refs/heads/main/attachments/23770/restBug%20%281%29.JPG)
  > ![restBug (1).JPG](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-core-cli/refs/heads/main/attachments/23770/restBug%20%281%29.JPG)
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jmiller.jpg" title="jmiller's avatar" /> jmiller:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-18407?focusedId=180508&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-180508">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I'm seeing strange behavior here as well. When starting a job with params via the UI, sometimes the result is a '201 CREATED' (and a blank/white page in the browser), and sometimes a 302 redirect back to the job page as in previous versions.</p>
</pre>
</details>

</ul>
</details>

<p>I'm seeing strange behavior here as well. When starting a job with params via the UI, sometimes the result is a '201 CREATED' (and a blank/white page in the browser), and sometimes a 302 redirect back to the job page as in previous versions.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kohsuke.png" title="kohsuke's avatar" /> kohsuke:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-18407?focusedId=180543&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-180543">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>To explain what's going on, 201 gets returned when the item gets submitted in the queue, and 302 if the item was already in the queue.</p>

<p>I realize that in both cases it probably makes sense to return 201, even though in the latter case, technically we haven't created any resources. Most often, the caller doesn't really care if the submission resulted in a new item in the queue or if it was already in the queue.</p>
</pre>
</details>

</ul>
</details>

<p>To explain what's going on, 201 gets returned when the item gets submitted in the queue, and 302 if the item was already in the queue.</p>

<p>I realize that in both cases it probably makes sense to return 201, even though in the latter case, technically we haven't created any resources. Most often, the caller doesn't really care if the submission resulted in a new item in the queue or if it was already in the queue.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/scm_issue_link.png" title="scm_issue_link's avatar" /> scm_issue_link:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-18407?focusedId=180546&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-180546">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Code changed in jenkins
User: Kohsuke Kawaguchi
Path:
 core/src/main/java/hudson/matrix/MatrixConfiguration.java
 core/src/main/java/hudson/matrix/listeners/MatrixBuildListener.java
 core/src/main/java/hudson/model/AbstractProject.java
 core/src/main/java/hudson/model/Queue.java
 core/src/main/java/hudson/model/queue/ScheduleResult.java
 test/src/test/groovy/hudson/model/AbstractProjectTest.groovy
<a href="http://jenkins-ci.org/commit/jenkins/f719bd06a46b0c8f8249f5d4848db61947b920be" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/jenkins/f719bd06a46b0c8f8249f5d4848db61947b920be</a>
Log:
  <span class="error">&#91;FIXED JENKINS-18407&#93;</span></p>

<p>Added Queue.schedule2 to allow the caller to retrieve the existing item in the queue. AbstractProject.doBuild() changed the behavior a bit to reply 201 if the item was already found in the queue (instead of a new one created.)</p>


<p>Compare: <a href="https://github.com/jenkinsci/jenkins/compare/bdfc3f715630...f719bd06a46b" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/jenkins/compare/bdfc3f715630...f719bd06a46b</a></p>

</pre>
</details>

</ul>
</details>

<p>Code changed in jenkins<br/>
User: Kohsuke Kawaguchi<br/>
Path:<br/>
 core/src/main/java/hudson/matrix/MatrixConfiguration.java<br/>
 core/src/main/java/hudson/matrix/listeners/MatrixBuildListener.java<br/>
 core/src/main/java/hudson/model/AbstractProject.java<br/>
 core/src/main/java/hudson/model/Queue.java<br/>
 core/src/main/java/hudson/model/queue/ScheduleResult.java<br/>
 test/src/test/groovy/hudson/model/AbstractProjectTest.groovy<br/>
<a href="http://jenkins-ci.org/commit/jenkins/f719bd06a46b0c8f8249f5d4848db61947b920be" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/jenkins/f719bd06a46b0c8f8249f5d4848db61947b920be</a><br/>
Log:<br/>
  <span class="error">&#91;FIXED JENKINS-18407&#93;</span></p>

<p>Added Queue.schedule2 to allow the caller to retrieve the existing item in the queue. AbstractProject.doBuild() changed the behavior a bit to reply 201 if the item was already found in the queue (instead of a new one created.)</p>


<p>Compare: <a href="https://github.com/jenkinsci/jenkins/compare/bdfc3f715630...f719bd06a46b" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/jenkins/compare/bdfc3f715630...f719bd06a46b</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/dogfood.png" title="dogfood's avatar" /> dogfood:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-18407?focusedId=180569&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-180569">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Integrated in <span class="image-wrap" style=""><img src="http://ci.jenkins-ci.org/images/16x16/yellow.png" style="border: 0px solid black" /></span> <a href="http://ci.jenkins-ci.org/job/jenkins_main_trunk/2661/" class="external-link" target="_blank" rel="nofollow noopener">jenkins_main_trunk #2661</a></p>

<p>     Result = UNSTABLE</p>
</pre>
</details>

</ul>
</details>

<p>Integrated in <span class="image-wrap" style=""><img src="http://ci.jenkins-ci.org/images/16x16/yellow.png" style="border: 0px solid black" /></span> <a href="http://ci.jenkins-ci.org/job/jenkins_main_trunk/2661/" class="external-link" target="_blank" rel="nofollow noopener">jenkins_main_trunk #2661</a></p>

<p>     Result = UNSTABLE</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/amoss166.png" title="amoss166's avatar" /> amoss166:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-18407?focusedId=180759&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-180759">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I used the latest build (not official one) in order to test the fix, and this time I'm getting HTTP CODE 200</p>

<p>I created a simple job with one string parameter and used the following code:</p>


<p>import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;</p>

<p>import org.codehaus.jackson.JsonFactory;
import org.codehaus.jackson.JsonNode;
import org.codehaus.jackson.JsonProcessingException;
import org.codehaus.jackson.map.ObjectMapper;
import org.springframework.beans.factory.InitializingBean;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpMethod;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Component;
import org.springframework.util.LinkedMultiValueMap;
import org.springframework.util.MultiValueMap;
import org.springframework.web.client.RestTemplate;</p>

<p>.
.
.
HttpHeaders headers = new HttpHeaders();
headers.setContentType(MediaType.APPLICATION_FORM_URLENCODED);</p>

<p>MultiValueMap&lt;String, String&gt; form = new LinkedMultiValueMap&lt;String, String&gt;();
form.add("COUNT", "5");</p>

<p>HttpEntity&lt;MultiValueMap&lt;String, String&gt;&gt; request = new HttpEntity&lt;MultiValueMap&lt;String, String&gt;&gt;(form,headers);</p>

<p>ResponseEntity&lt;String&gt; responce = restTemplate.exchange("http://192.168.1.126:8080/jenkins/job/test1/buildWithParameters", HttpMethod.POST, request, String.class);
.
.
.</p>

<p>The returned HTTP code is 200 instead of 201 (saw it on Fiddler)</p>

<p>Additionally, I'm getting a json result which contains the "queueItem" section, together with other non-relevant info, I can use this "queueItem" info for tracking the job, BUT when starting few instances of the same job in parallel (each instance with different parameters), in each HTTP response I'm getting the same "queueItem" as the first one I got, although, when checking the queue the other new instance were created with different item id in the queue</p>

<p>If this too complicated to understand, lets try to set a webex session or something similar (my mail: amos_son@malam.com)</p>



</pre>
</details>

</ul>
</details>

<p>I used the latest build (not official one) in order to test the fix, and this time I'm getting HTTP CODE 200</p>

<p>I created a simple job with one string parameter and used the following code:</p>


<p>import java.io.IOException;<br/>
import java.util.ArrayList;<br/>
import java.util.HashMap;<br/>
import java.util.List;<br/>
import java.util.Map;<br/>
import java.util.Map.Entry;</p>

<p>import org.codehaus.jackson.JsonFactory;<br/>
import org.codehaus.jackson.JsonNode;<br/>
import org.codehaus.jackson.JsonProcessingException;<br/>
import org.codehaus.jackson.map.ObjectMapper;<br/>
import org.springframework.beans.factory.InitializingBean;<br/>
import org.springframework.beans.factory.annotation.Autowired;<br/>
import org.springframework.http.HttpEntity;<br/>
import org.springframework.http.HttpHeaders;<br/>
import org.springframework.http.HttpMethod;<br/>
import org.springframework.http.HttpStatus;<br/>
import org.springframework.http.MediaType;<br/>
import org.springframework.http.ResponseEntity;<br/>
import org.springframework.stereotype.Component;<br/>
import org.springframework.util.LinkedMultiValueMap;<br/>
import org.springframework.util.MultiValueMap;<br/>
import org.springframework.web.client.RestTemplate;</p>

<p>.<br/>
.<br/>
.<br/>
HttpHeaders headers = new HttpHeaders();<br/>
headers.setContentType(MediaType.APPLICATION_FORM_URLENCODED);</p>

<p>MultiValueMap<String, String> form = new LinkedMultiValueMap<String, String>();<br/>
form.add("COUNT", "5");</p>

<p>HttpEntity<MultiValueMap<String, String>> request = new HttpEntity<MultiValueMap<String, String>>(form,headers);</p>

<p>ResponseEntity<String> responce = restTemplate.exchange("http://192.168.1.126:8080/jenkins/job/test1/buildWithParameters", HttpMethod.POST, request, String.class);<br/>
.<br/>
.<br/>
.</p>

<p>The returned HTTP code is 200 instead of 201 (saw it on Fiddler)</p>

<p>Additionally, I'm getting a json result which contains the "queueItem" section, together with other non-relevant info, I can use this "queueItem" info for tracking the job, BUT when starting few instances of the same job in parallel (each instance with different parameters), in each HTTP response I'm getting the same "queueItem" as the first one I got, although, when checking the queue the other new instance were created with different item id in the queue</p>

<p>If this too complicated to understand, lets try to set a webex session or something similar (my mail: amos_son@​malam.com)</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>barthelemy:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-18407?focusedId=192402&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-192402">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I confirm that using the REST API, I get a 200 instead of a 201 when performing a build. The URL I call is :
https://&lt;hidden&gt;/jenkins/job/&lt;JOB&gt;/buildWithParameters?module=amoreTST</p>

</pre>
</details>

</ul>
</details>

<p>I confirm that using the REST API, I get a 200 instead of a 201 when performing a build. The URL I call is :<br/>
https://<hidden>/jenkins/job/<JOB>/buildWithParameters?module=amoreTST</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>mariuszs:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-18407?focusedId=195006&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-195006">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>More importantly, buildWithParameters returns 201 without Location to build queue.</p>

<p>&gt; the successful queueing will result in 201 status code with Location HTTP header pointing the URL of the item in the queue. </p>
</pre>
</details>

</ul>
</details>

<p>More importantly, buildWithParameters returns 201 without Location to build queue.</p>

<p>> the successful queueing will result in 201 status code with Location HTTP header pointing the URL of the item in the queue. </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-18407?focusedId=199592&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-199592">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>The current version of <tt>ParametersDefinitionProperty.buildWithParameters</tt> does seem to send either a 201 with redirect to a newly created or existing queue item, or a 200 with redirect to the job in case queue scheduling was refused. If you can reproduce different behavior please file a pull request with accompanying test.</p>

<p>Remember that you need to either POST (with a valid crumb, if Jenkins is so configured), or (for historical compatibility) GET with a <tt>token</tt> (if one is defined), in order to schedule a build: for security reasons, GET requests may not initiate actions.</p>
</pre>
</details>

</ul>
</details>

<p>The current version of <tt>ParametersDefinitionProperty.buildWithParameters</tt> does seem to send either a 201 with redirect to a newly created or existing queue item, or a 200 with redirect to the job in case queue scheduling was refused. If you can reproduce different behavior please file a pull request with accompanying test.</p>

<p>Remember that you need to either POST (with a valid crumb, if Jenkins is so configured), or (for historical compatibility) GET with a <tt>token</tt> (if one is defined), in order to schedule a build: for security reasons, GET requests may not initiate actions.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/scm_issue_link.png" title="scm_issue_link's avatar" /> scm_issue_link:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-18407?focusedId=203182&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-203182">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Code changed in jenkins
User: Jesse Glick
Path:
 src/test/java/hudson/plugins/git/RevisionParameterActionTest.java
<a href="http://jenkins-ci.org/commit/git-plugin/4fe33a1fa31459563b7d9e4c4c652b20bde6df4d" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/git-plugin/4fe33a1fa31459563b7d9e4c4c652b20bde6df4d</a>
Log:
  Commenting out assertions that fail in 1.521+ due to improved core behavior from <a href="https://issues.jenkins.io/browse/JENKINS-18407" title="Starting a job with params using REST Api doesn't return 201 http status code as stated in the documentation" class="issue-link" data-issue-key="JENKINS-18407"><del>JENKINS-18407</del></a>.</p>


</pre>
</details>

</ul>
</details>

<p>Code changed in jenkins<br/>
User: Jesse Glick<br/>
Path:<br/>
 src/test/java/hudson/plugins/git/RevisionParameterActionTest.java<br/>
<a href="http://jenkins-ci.org/commit/git-plugin/4fe33a1fa31459563b7d9e4c4c652b20bde6df4d" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/git-plugin/4fe33a1fa31459563b7d9e4c4c652b20bde6df4d</a><br/>
Log:<br/>
  Commenting out assertions that fail in 1.521+ due to improved core behavior from <a href="https://issues.jenkins.io/browse/JENKINS-18407" title="Starting a job with params using REST Api doesn't return 201 http status code as stated in the documentation" class="issue-link" data-issue-key="JENKINS-18407"><del>JENKINS-18407</del></a>.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/scm_issue_link.png" title="scm_issue_link's avatar" /> scm_issue_link:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-18407?focusedId=256318&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-256318">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Code changed in jenkins
User: Craig Rodrigues
Path:
 src/test/groovy/com/cloudbees/plugins/flow/ConcurrencyTest.groovy
<a href="http://jenkins-ci.org/commit/build-flow-plugin/faeaa5cc2062f7c7b73073e201b12646e9538963" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/build-flow-plugin/faeaa5cc2062f7c7b73073e201b12646e9538963</a>
Log:
  Delete test which no longer works in 1.521+ due to improved core behavior from <a href="https://issues.jenkins.io/browse/JENKINS-18407" title="Starting a job with params using REST Api doesn't return 201 http status code as stated in the documentation" class="issue-link" data-issue-key="JENKINS-18407"><del>JENKINS-18407</del></a></p>

<p>Due to the improved core behavior, in JobInvocation.groovy's run()
method, schedule2() no longer returns null when invoked from the test.</p>


</pre>
</details>

</ul>
</details>

<p>Code changed in jenkins<br/>
User: Craig Rodrigues<br/>
Path:<br/>
 src/test/groovy/com/cloudbees/plugins/flow/ConcurrencyTest.groovy<br/>
<a href="http://jenkins-ci.org/commit/build-flow-plugin/faeaa5cc2062f7c7b73073e201b12646e9538963" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/build-flow-plugin/faeaa5cc2062f7c7b73073e201b12646e9538963</a><br/>
Log:<br/>
  Delete test which no longer works in 1.521+ due to improved core behavior from <a href="https://issues.jenkins.io/browse/JENKINS-18407" title="Starting a job with params using REST Api doesn't return 201 http status code as stated in the documentation" class="issue-link" data-issue-key="JENKINS-18407"><del>JENKINS-18407</del></a></p>

<p>Due to the improved core behavior, in JobInvocation.groovy's run()<br/>
method, schedule2() no longer returns null when invoked from the test.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jnazander.png" title="jnazander's avatar" /> jnazander:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-18407?focusedId=350429&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-350429">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>It seems that this bug is still present (version 2.138.1) - every time when a build is triggered using `POST /job/<span class="error">&#91;JOB_NAME&#93;</span>/build` (either normally or via Curl), jenkins ALWAYS returns 302 instead of 201. This makes it impossible to trigger builds via a webhook, because the HTTP client sees that the response code is not 2xx and starts retrying it.</p>
</pre>
</details>

</ul>
</details>

<p>It seems that this bug is still present (version 2.138.1) - every time when a build is triggered using `POST /job/<span class="error">&#91;JOB_NAME&#93;</span>/build` (either normally or via Curl), jenkins ALWAYS returns 302 instead of 201. This makes it impossible to trigger builds via a webhook, because the HTTP client sees that the response code is not 2xx and starts retrying it.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-18407?focusedId=350436&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-350436">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jnazander" class="user-hover" rel="jnazander">jnazander</a> no you are just using the wrong endpoint (<tt>build</tt> vs. <tt>buildWithParameters</tt>). Please use the users’ list for assistance. If you believe you have encountered a bug please file a fresh report with complete steps to reproduce from scratch, linking to existing issues as appropriate.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jnazander" class="user-hover" rel="jnazander">jnazander</a> no you are just using the wrong endpoint (<tt>build</tt> vs. <tt>buildWithParameters</tt>). Please use the users’ list for assistance. If you believe you have encountered a bug please file a fresh report with complete steps to reproduce from scratch, linking to existing issues as appropriate.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jnazander.png" title="jnazander's avatar" /> jnazander:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-18407?focusedId=350442&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-350442">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick" class="user-hover" rel="jglick">jglick</a> But my job doesn't have parameters, and if I try calling the job using `/buildWithParameters`, it returns HTTP 500 – `IllegalStateException: This build is not parameterized!`</p>

<p> </p>

<p>In any case, Jenkins's API documentation says:</p>

<p>`To programmatically schedule a new build, post to this URL (<a href="https://jenkins.***.net/job/accounts/build" class="external-link" target="_blank" rel="nofollow noopener">https://jenkins.***.net/job/accounts/build</a>). If the build has parameters, post to this URL and provide the parameters as form data. Either way, the successful queueing will result in 201 status code`</p>

<p>While in reality it returns 302. This looks like a bug to me, should I create a new ticket?</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick" class="user-hover" rel="jglick">jglick</a> But my job doesn't have parameters, and if I try calling the job using `/buildWithParameters`, it returns HTTP 500 – `IllegalStateException: This build is not parameterized!`</p>

<p> </p>

<p>In any case, Jenkins's API documentation says:</p>

<p>`To programmatically schedule a new build, post to this URL (<a href="https://jenkins.***.net/job/accounts/build" class="external-link" target="_blank" rel="nofollow noopener">https://jenkins.***.net/job/accounts/build</a>). If the build has parameters, post to this URL and provide the parameters as form data. Either way, the successful queueing will result in 201 status code`</p>

<p>While in reality it returns 302. This looks like a bug to me, should I create a new ticket?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-18407?focusedId=350447&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-350447">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>If your job has parameters, use <tt>buildWithParameters</tt>. If it does not, use <tt>build</tt>. Which is what the API documentation says. There is no bug I can see. Again please use the users’ list for assistance.</p>
</pre>
</details>

</ul>
</details>

<p>If your job has parameters, use <tt>buildWithParameters</tt>. If it does not, use <tt>build</tt>. Which is what the API documentation says. There is no bug I can see. Again please use the users’ list for assistance.</p>
